Build failure: docfd

### Steps To Reproduce

nix shell nixpkgs#docfd

### Build log

```
error: builder for '/nix/store/pw9wpp7v2mkp9ijnqwpn12hv6b2n2y6k-ocaml5.1.1-eio_posix-1.0.drv' failed with exit code 1;
       > eio_posix_stubs.c:133:7: error: call to undeclared function 'preadv'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
       >   r = preadv(Int_val(v_fd), iov, n_bufs, offset);
       >       ^
       > eio_posix_stubs.c:150:7: error: call to undeclared function 'pwritev'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
       >   r = pwritev(Int_val(v_fd), iov, n_bufs, offset);
       >       ^
       > 2 errors generated.
       >
       For full logs, run 'nix log /nix/store/pw9wpp7v2mkp9ijnqwpn12hv6b2n2y6k-ocaml5.1.1-eio_posix-1.0.drv'.
error: 1 dependencies of derivation '/nix/store/1b9x6a5fj12cmp1prwkyk803g467wknz-ocaml5.1.1-eio_main-1.0.drv' failed to build
error: 1 dependencies of derivation '/nix/store/gly8zm9w97ffh35xv2dhzy73d0b3gcxz-ocaml5.1.1-docfd-3.0.0.drv' failed to build
```

### Additional context

Wanted to try out `docfd` but it failed to build. I noticed the package is not on the [latest version](https://github.com/darrenldl/docfd/releases/tag/4.0.0).

### Notify maintainers

@chewblacka 

### Metadata

```console
 - system: `"x86_64-darwin"`
 - host os: `Darwin 21.6.0, macOS 10.16`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - channels(root): `""`
 - nixpkgs: `not found`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/do/docfd/package.nix
+++ b/pkgs/by-name/do/docfd/package.nix
@@ -1,5 +1,7 @@
 { lib
 , ocamlPackages
+, stdenv
+, overrideSDK
 , fetchFromGitHub
 , python3
 , dune_3
@@ -10,7 +12,13 @@
 , docfd
 }:
 
-ocamlPackages.buildDunePackage rec {
+let
+  # Needed for x86_64-darwin
+  buildDunePackage' = ocamlPackages.buildDunePackage.override {
+    stdenv = if stdenv.isDarwin then overrideSDK stdenv "11.0" else stdenv;
+  };
+in
+buildDunePackage' rec {
   pname = "docfd";
   version = "4.0.0";
 

--- a/pkgs/by-name/do/docfd/package.nix
+++ b/pkgs/by-name/do/docfd/package.nix
@@ -3,6 +3,8 @@
 , fetchFromGitHub
 , python3
 , dune_3
+, testers
+, docfd
 }:
 
 ocamlPackages.buildDunePackage rec {
@@ -34,6 +36,10 @@ ocamlPackages.buildDunePackage rec {
     timedesc
     yojson
   ];
+
+  passthru.tests.version = testers.testVersion {
+    package = docfd;
+  };
 
   meta = with lib; {
     description = "TUI multiline fuzzy document finder";

--- a/pkgs/by-name/do/docfd/package.nix
+++ b/pkgs/by-name/do/docfd/package.nix
@@ -3,13 +3,16 @@
 , fetchFromGitHub
 , python3
 , dune_3
+, makeWrapper
+, pandoc
+, poppler_utils
 , testers
 , docfd
 }:
 
 ocamlPackages.buildDunePackage rec {
   pname = "docfd";
-  version = "3.0.0";
+  version = "4.0.0";
 
   minimalOCamlVersion = "5.1";
 
@@ -17,10 +20,10 @@ ocamlPackages.buildDunePackage rec {
     owner = "darrenldl";
     repo = "docfd";
     rev = version;
-    hash = "sha256-pJ5LlOfC+9NRfY7ng9LAxEnjr+mtJmhRNTo9Im6Lkbo=";
+    hash = "sha256-fgwUXRZ6k5i3XLxXpjbrl0TJZMT+NkGXf7KNwRgi+q8=";
   };
 
-  nativeBuildInputs = [ python3 dune_3 ];
+  nativeBuildInputs = [ python3 dune_3 makeWrapper ];
   buildInputs = with ocamlPackages; [
     cmdliner
     containers-data
@@ -37,6 +40,10 @@ ocamlPackages.buildDunePackage rec {
     yojson
   ];
 
+  postInstall = ''
+    wrapProgram $out/bin/docfd --prefix PATH : "${lib.makeBinPath [ pandoc poppler_utils ]}"
+  '';
+
   passthru.tests.version = testers.testVersion {
     package = docfd;
   };
@@ -44,14 +51,10 @@ ocamlPackages.buildDunePackage rec {
   meta = with lib; {
     description = "TUI multiline fuzzy document finder";
     longDescription = ''
-      Think interactive grep for both text and other document files, but
-      word/token based instead of regex and line based, so you can search
-      across lines easily. Aims to provide good UX via integration with
-      common text editors and other file viewers.
-      Optional dependencies:
-        fzf - for fuzzy file picker with "docfd ?".
-        poppler_utils - for pdf search.
-        pandoc - for .epub, .odt, .docx, .fb2, .ipynb, .html, & .htm files.
+      Think interactive grep for text and other document files.
+      Word/token based instead of regex and line based, so you
+      can search across lines easily. Aims to provide good UX via
+      integration with common text editors and other file viewers.
     '';
     homepage = "https://github.com/darrenldl/docfd";
     license = licenses.mit;

--- a/pkgs/development/ocaml-modules/eio/posix.nix
+++ b/pkgs/development/ocaml-modules/eio/posix.nix
@@ -1,4 +1,6 @@
 { buildDunePackage
+, lib
+, stdenv
 , dune-configurator
 , eio
 , fmt
@@ -13,6 +15,10 @@ buildDunePackage {
   minimalOCamlVersion = "5.0";
 
   dontStrip = true;
+
+  env = lib.optionalAttrs stdenv.isDarwin {
+    NIX_CFLAGS_COMPILE = "-Wno-implicit-function-declaration";
+  };
 
   buildInputs = [
     dune-configurator

